plugins/ep963x/fu-ep963x-device.c

```diff
@@ -0,0 +1,347 @@
+				       FU_EP963_DEVICE_TIMEOUT,
+				       FU_HID_DEVICE_FLAG_IS_FEATURE,
+				       error)) {
+		return FALSE;
+	}
+
+	/* wait for hardware */
+	for (guint i = 0; i < 16; i++) {
+		g_usleep (100 * 1000);
+		bufhw[0] = FU_EP963_FEATURE_ID1_SIZE;
+		bufhw[1] = cmd;
+		if (!fu_hid_device_get_report (FU_HID_DEVICE (self), 0x00,
+					       bufhw, sizeof(bufhw),
+					       FU_EP963_DEVICE_TIMEOUT,
+					       FU_HID_DEVICE_FLAG_IS_FEATURE,
+					       error)) {
+			return FALSE;
+		}
+		if (bufhw[0x07] != FU_EP963_SMBUS_ERROR_NONE) {
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_WRITE,
+				     "smbus failed %s",
+				     fu_ep963x_smbus_strerror (bufhw[0x07]));
+			return FALSE;
+		}
+		if (bufhw[0x02] == FU_EP963_USB_STATE_READY)
+			return TRUE;
+
+		/* USB state changed, so reset counter */
+		if (usb_state != bufhw[0x02]) {
+			usb_state = bufhw[0x02];
+			i = 0;
+		}
+	}
+
+	/* failed */
+	g_set_error_literal (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_WRITE,
+			     "failed to wait for icp-done");
+	return FALSE;
+}
+
+static gboolean
+fu_ep963x_device_detach (FuDevice *device, GError **error)
+{
+	FuEp963xDevice *self = FU_EP963X_DEVICE (device);
+	const guint8 buf[] = { 'E', 'P', '9', '6', '3' };
+	g_autoptr(GError) error_local = NULL;
+
+	if (!fu_ep963x_device_write_icp (self, FU_EP963_OPCODE_SUBMCU_ENTER_ICP,
+					 buf, sizeof(buf),
+					 &error_local)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_WRITE,
+			     "failed to detach: %s",
+			     error_local->message);
+		return FALSE;
+	}
+
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	return TRUE;
+}
+
+static gboolean
+fu_ep963x_device_attach (FuDevice *device, GError **error)
+{
+	FuEp963xDevice *self = FU_EP963X_DEVICE (device);
+	const guint8 buf[] = { 0x00 };
+	g_autoptr(GError) error_local = NULL;
+
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+	if (!fu_ep963x_device_write_icp (self, 0x02,
+					 buf, sizeof(buf),
+					 &error_local)) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_WRITE,
+			     "failed to boot to runtime: %s",
+			     error_local->message);
+		return FALSE;
+	}
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	return TRUE;
+}
+
+static gboolean
+fu_ep963x_device_setup (FuDevice *device, GError **error)
+{
+	FuEp963xDevice *self = FU_EP963X_DEVICE (device);
+	guint8 buf[FU_EP963_FEATURE_ID1_SIZE - 2] = { 0x0 };
+	g_autofree gchar *version = NULL;
+
+	/* get version */
+	if (!fu_ep963x_device_write_icp (self, FU_EP963_UF_CMD_VERSION,
+					 buf, sizeof(buf),
+					 error)) {
+		return FALSE;
+	}
+	version = g_strndup ((const gchar *) buf, 5);
+	fu_device_set_version (device, version);
+
+	/* the VID and PID are unchanged between bootloader modes */
+	if (g_strcmp0 (version, "00000") == 0) {
+		fu_device_set_flags (device, FWUPD_DEVICE_FLAG_IS_BOOTLOADER);
```

Comment:
I would say you should add instead of set to prevent debugging an error later

Reply:
Doh! good catch!